siege_utilities/reporting/report_generator.py

```
@@ -454,11 +498,30 @@ def generate_pdf_report(self, report_content: Dict[str, Any],
             template.build_document(story)
 
+            # Atomic rename. os.replace is atomic on POSIX and on Windows
+            # (per docs) when source and dest live on the same FS — which
+            # they do, since we created the temp alongside the target.
+            try:
+                os.replace(tmp_path, final)
+            except OSError as exc:
+                log.error(
+                    "generate_pdf_report: built %s but could not rename "
+                    "to %s: %s", tmp_path, final, exc,
+                )
```

Ruff (0.15.12)

[warning] Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)